README
= Radix

* http://death.rubyforge.org
* http://death.rubyforge.org/radix


== DESCRIPTION

Radix provides the means of converting to and from any base.
In addition, representational notations need not be in
ASCII order; any user-defined notation can be used upto
base 62.


== FEATURES/ISSUES

* Convert to and from any base.
* User-definable character set upto base 62.
* Defaults to standard base 62.
* Can be used to encode strings.


== RELEASE NOTES

Please see HISTORY file.


== SYNOPSIS

Base conversions with ASCII ordered notations are easy in Ruby.

  255.to_s(16)   #=> "FF"
  "FF".to_i(16)  #=> 255

But Ruby reaches it's limit at base 36.

  255.to_s(37)   #=> Error

Radix provides the means of converting to and from any base.

  Radix.convert_base([100, 10], 256, 10)
  #=> [2,5,6,1,0]

And it can handle any notation upto base 62.

  Radix.convert("10", 62, 10)  #=> "62"

And the notations need not be in ASCII order --odd notations
can be used.

  b10 = Radix.new([:Q, :W, :E, :R, :T, :Y, :U, :I, :O, :U])
  b10.convert("FF", 16) #=> "EYY"


== HOW TO INSTALL

To install with RubyGems simply open a console and type:

  gem install radix

Site installation requires Setup.rb (gem install setup),
then download the tarball package and type:

  tar -xvzf radix-1.0.0.tgz
  cd radix-1.0.0
  sudo setup.rb all

Windows users use 'ruby setup.rb all'.


== LINCENSE/COPYRIGHT

Copyright (c) 2006 The Coding Dead <http://death.rubyforge.org>

This program is ditributed unser the terms of the LGPLv3 license.

See LICENSE file for details.


HISTORY
= Release History

== 1.1.0 / 2010-09-03

Radix now provides an actual Numeric subclass, Radix::Number, that stores
the base and can be used like any other Numeric object. This makes it very
easy to convert and manipulate numbers in any base. The implementation is still
a bit nascent. For the moment, it only supports the most basic math operators
and only handles integer values, but furture releases will continue to expand
on it's capabilites.

Changes:

* Add Radix::Number to handle bases like any other numeric.
* Rename Radix class to Radix::Base.


== 1.0.0 / 2009-07-01

This is the initial stand-alone release of Radix,
ported from the basex.rb library of Ruby Facets.

Changes:

* Happy Birthday!



lib/radix/base.rb
# Radix coverts to and from any base.
#
# Base conversions with ASCII ordered notations are easy in Ruby.
#
#   255.to_s(16)   #=> "FF"
#   "FF".to_i(16)  #=> 255
#
# But Ruby reaches it's limit at base 36.
#
#   255.to_s(37)   #=> Error
#
# Radix provides the means of converting to and from any base.
#
#   Radix::Base.convert_base([100, 10], 256, 10)
#   #=> [2,5,6,1,0]
#
# And it can handle any notation upto base 62.
#
#   Radix::Base.convert("10", 62, 10)  #=> "62"
#
# And the notations need not be in ASCII order --odd notations
# can be used.
#
#   b10 = Radix::Base.new([:Q, :W, :E, :R, :T, :Y, :U, :I, :O, :U])
#   b10.convert("FF", 16) #=> "EYY"
#
module Radix

  class Base

    BASE10 = ["0".."9"].map { |r| r.to_a }.flatten
    BASE12 = ["0".."9", ["X", "E"]].map { |r| r.to_a }.flatten
    BASE16 = ["0".."9", "A".."F"].map { |r| r.to_a }.flatten
    BASE36 = ["0".."9", "A".."Z"].map { |r| r.to_a }.flatten
    BASE60 = ["0".."9", "a".."z", "A".."X"].map { |r| r.to_a }.flatten
    BASE62 = ["0".."9", "a".."z", "A".."Z"].map { |r| r.to_a }.flatten

    attr :chars
    attr :base
    attr :values

    # New Radix using +chars+ representation.
    def initialize(chars=BASE62)
      @chars  = chars.map{ |c| c.to_s }
      @base   = @chars.size
      @values = Hash[*(0...@base).map { |i| [ @chars[i], i ] }.flatten]
    end

    # Encode a string in the radix.
    def encode(byte_string)
      digits = byte_string.unpack("C*")
      digits = convert_base(digits, 256, base)
      digits.map{ |d| @chars[d] }.join
    end

    # Decode a string that was previously encoded in the radix.
    def decode(encoded)
      digits = encoded.split(//).map{ |c| @values[c] }
      convert_base(digits, base, 256).pack("C*")
    end

    # Convert a representational +number+ of +from_radix+ to the radix.
    def convert(number, from_radix)
      from_radix = standard_radix(from_radix) if Integer === from_radix
      digits = number.to_s.split(//)
      digits = digits.map{ |digit| from_radix.values[digit] }
      digits = convert_base(digits, from_radix.base, base)
      digits = digits.map{ |digit| chars[digit] }
      digits.join
    end

    # Convert any base to any other base, using array of +digits+.
    def convert_base(digits, from_base, to_base)
      self.class.convert_base(digits, from_base, to_base)
    end

    private

    def standard_radix(integer_base)
      self.class.standard_radix(integer_base)
    end

    public

    # Do a standard conversion upto base 62.
    def self.convert(number, from_base, to_base)
      r1 = standard_radix(from_base)
      r2 = standard_radix(to_base)
      r2.convert(number, r1)
    end

    # Convert any base to any other base, using array of +digits+.
    def self.convert_base(digits, from_base, to_base)
      bignum = 0
      digits.each { |digit| bignum = bignum * from_base + digit }
      converted = []
      until bignum.zero?
        bignum, digit = bignum.divmod(to_base)
        converted.push(digit)
      end
      converted.reverse
    end

    # Provide a standard representation of a base upto 62.
    def self.standard_radix(integer_base)
      if integer_base > 36
        new(BASE62[0..integer_base-1])
      else
        new(BASE36[0..integer_base-1])
      end
    end

  end

  # Do a standard conversion upto base 62.
  def self.convert(number, from_base, to_base)
    Radix::Base.convert(number, from_base, to_base)
  end

  # Convert any base to any other base, using array of +digits+.
  def self.convert_base(digits, from_base, to_base)
    Radix::Base.convert_base(digits, from_base, to_base)
  end

  # Provide a standard representation of a base upto 62.
  def self.standard_radix(integer_base)
    Radix::Base.standard_radix(integer_base)
  end

end

